File: row-change-event-channel/src/test/scala/nl/haploid/event/channel/ContainerizedPostgresConfiguration.scala
package nl.haploid.event.channel

import java.nio.file.Files
import javax.sql.DataSource

import com.github.dockerjava.api.DockerClient
import com.github.dockerjava.api.model._
import com.jolbox.bonecp.BoneCPDataSource
import grizzled.slf4j.Logging
import org.springframework.beans.factory.annotation.{Autowired, Value}
import org.springframework.context.annotation.{Bean, Configuration}

import scala.collection.JavaConverters._

@Configuration
class ContainerizedPostgresConfiguration extends PostgresConfiguration with Logging {

  val IMAGE_NAME = "haploid/postgres:latest"
  val CONTAINER_NAME: String = "postgres_it"

  @Value("${DOCKER_HOST}") val dockerHost: String = null

  @Autowired val docker: DockerClient = null

  override val port: Int = 5432
  override val username = "postgres"
  override val database = "postgres"

  override def hostname = dockerHost.replaceAll("^.*?(\\d+[.]\\d+[.]\\d+[.]\\d+).*?$", "$1")

  def startContainer = {
    val exposedPort = ExposedPort.tcp(port)
    val container = docker.createContainerCmd(IMAGE_NAME)
      .withName(CONTAINER_NAME)
      .withExposedPorts(exposedPort)
      .exec
    val portBindings = new Ports(exposedPort, Ports.Binding(port))
    val tempDir = Files.createTempDirectory("postgres").toFile.getAbsolutePath
    docker.startContainerCmd(container.getId)
      .withBinds(new Bind(tempDir, new Volume("/var/lib/postgresql/data")))
      .withPortBindings(portBindings)
      .exec
    docker.waitContainerCmd(container.getId)
    Thread.sleep(5000)
    container
  }

  def stopContainer = {
    val containers = docker
      .listContainersCmd
      .withShowAll(true)
      .exec
      .asScala
      .filter(c => c.getNames.contains("/%s".format(CONTAINER_NAME)))
    containers
      .filter(c => docker.inspectContainerCmd(c.getId).exec.getState.isRunning)
      .foreach(c => docker.stopContainerCmd(c.getId).exec)
    containers.foreach(c => docker.removeContainerCmd(c.getId).exec)
    containers
  }

  @Bean
  override def dataSource: DataSource = {

    stopContainer
    startContainer

    val dataSource = new BoneCPDataSource
    dataSource.setDriverClass("org.postgresql.Driver")
    val jdbcUrl = "jdbc:postgresql://%s:%d/%s".format(hostname, port, database)
    dataSource.setJdbcUrl(jdbcUrl)
    dataSource.setUsername(username)
    dataSource.setPassword(username)
    debug("Will connect to %s as %s".format(jdbcUrl, username))
    dataSource
  }
}

File: row-change-event-channel/src/test/java/nl/haploid/event/channel/ContainerizedPostgresConfiguration.java
package nl.haploid.event.channel;

import com.github.dockerjava.api.model.Bind;
import com.github.dockerjava.api.model.ExposedPort;
import com.github.dockerjava.api.model.Ports;
import com.github.dockerjava.api.model.Volume;
import com.jolbox.bonecp.BoneCPDataSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;

import javax.sql.DataSource;

@Configuration
@ComponentScan // TODO: okay, it's a little annoying to do a component scan here
public class ContainerizedPostgresConfiguration extends PostgresConfiguration {

    private static final int PORT = 5432;
    private static final String USERNAME = "postgres";
    private static final String DATABASE = "postgres";
    private static final String IMAGE_NAME = "postgres:9.2";
    private static final String CONTAINER_NAME = "postgres_it";

    private final Logger log = LoggerFactory.getLogger(getClass());

    @Value("${DOCKER_HOST}")
    private String dockerHost;

    @Autowired
    private DockerService dockerService;

    @Override
    public String getHostname() {
        return dockerHost.replaceAll("^.*?(\\d+[.]\\d+[.]\\d+[.]\\d+).*?$", "$1");
    }

    private void waitForPostgresToStart() {
        try {
            Thread.sleep(5000);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    @Bean
    @Override
    public DataSource dataSource() {

        dockerService.stopContainer(CONTAINER_NAME);
        final Ports ports = new Ports(ExposedPort.tcp(PORT), Ports.Binding(PORT));
        final Bind bind = new Bind(dockerService.createTempDirectory(), new Volume("/var/lib/postgresql/data"));
        dockerService.startContainer(IMAGE_NAME, CONTAINER_NAME, ports, bind);
        waitForPostgresToStart();

        final BoneCPDataSource dataSource = new BoneCPDataSource();
        dataSource.setDriverClass("org.postgresql.Driver");
        final String jdbcUrl = String.format("jdbc:postgresql://%s:%d/%s", getHostname(), PORT, DATABASE);
        dataSource.setJdbcUrl(jdbcUrl);
        dataSource.setUsername(USERNAME);
        dataSource.setPassword(USERNAME);
        log.debug(String.format("Will connect to %s as %s", jdbcUrl, USERNAME));
        return dataSource;
    }
}
